pkg/v2/tkr/webhook/cluster/vsphere-template-resolver/template/resolver.go

```diff
@@ -0,0 +1,355 @@
+					return errors.Wrapf(err, "error setting control plane TKR_DATA variable")
+				}
+			}
+			cw.Log.Info("template resolution result processing complete for Control Plane topology", "version", topologyVersion)
+		}
+	}
+
+	// Process MD results
+	if len(mdDatas) == 0 {
+		return nil
+	}
+	for k := range mdDatas {
+		// Get the 'k'th data, this will be used to get the query and the TKR_DATA so we don't have to parse variables all over again.
+		mdData := mdDatas[k]
+		if mdData == nil {
+			continue
+		}
+
+		tkrData := *mdData.TKRData
+		tkrDataValue, ok := (tkrData)[topologyVersion]
+		if !ok {
+			// No matching entry for topology version, can skip.
+			continue
+		}
+
+		templateResult := (*result.MachineDeployments)[*mdData.TemplateQuery]
+		populateTKRDataFromResult(tkrDataValue, templateResult)
+		if err := util_topology.SetMDVariable(cluster, k, varTKRData, tkrData); err != nil {
+			return errors.Wrapf(err, fmt.Sprintf("error setting machine deployment [%v]'s TKR_DATA variable", k))
+		}
+	}
+	cw.Log.Info("template resolution result processing complete for machine deployments in topology", "version", topologyVersion)
+	return nil
+}
+
+func populateTKRDataFromResult(tkrDataValue *resolver_cluster.TKRDataValue, templateResult *templateresolver.TemplateResult) {
+	if templateResult == nil {
+		// If the values are empty, its possible that the resolution was skipped because the details were already present.
+		// Do not overwrite.
+		return
+	}
+	tkrDataValue.OSImageRef[keyOSImageTemplate] = templateResult.TemplatePath
+	tkrDataValue.OSImageRef[keyOSImageMOID] = templateResult.TemplateMOID
+}
+
+func (cw *Webhook) getVSphereContext(ctx context.Context, cluster *clusterv1.Cluster) (templateresolver.VSphereContext, error) {
+	c := cw.Client
+	// Get Secret (cluster name in cluster namespace)
+	clusterName, clusterNamespace := cluster.Name, cluster.Namespace
+
+	secret := &corev1.Secret{}
+	objKey := client.ObjectKey{Name: clusterName, Namespace: clusterNamespace}
+	cw.Log.Info("Getting secret for VC Client", "key", objKey)
+	err := c.Get(ctx, objKey, secret)
+	if err != nil {
+		return templateresolver.VSphereContext{}, errors.Wrap(err, fmt.Sprintf("could not get secret for key: %v", objKey))
+	}
+	username, password := secret.Data["username"], secret.Data["password"]
+
+	var vcClusterVar VCenterClusterVar
+	err = util_topology.GetVariable(cluster, varVCenter, &vcClusterVar)
+	if err != nil {
+		return templateresolver.VSphereContext{}, errors.Wrapf(err, "error parsing vcenter cluster variable")
+	}
+
+	insecure := true
+	if vcClusterVar.TLSThumbprint != "" {
+		insecure = false
+	}
+
+	// TODO(reviewers) / TODO(shashank): Is it ok to log this info below?
+	cw.Log.Info("Successfully retrieved vSphere context", "Server", vcClusterVar.Server, "datacenter", vcClusterVar.DataCenter, "tlsthumbprint", vcClusterVar.TLSThumbprint)
+	return templateresolver.VSphereContext{
```

Comment:
I think it should be fine to log these details.

Reply:
Thanks!